### PR TITLE
[SPARK-48385][SQL] Migrate the jdbc driver of mariadb from `2.x` to `3.x`

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to 
- migrate the jdbc driver of `mariadb` from `2.x` to `3.x`.
- upgrade the image version of `mariadb` from `10.5.12` to `10.5.24`.

### Why are the changes needed?
1.First PR of Series `3.x`:
https://github.com/mariadb-corporation/mariadb-connector-j/releases/tag/3.0.0-alpha
image

2.Seen from the Maven central repo, series `3. x` should have become the mainstream, with faster release frequency
https://mvnrepository.com/artifact/org.mariadb.jdbc/mariadb-java-client

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
- Update existed UT.
- Pass GA

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala [hunk 1]
```diff
@@ -46,7 +46,7 @@ class MariaDBKrbIntegrationSuite extends DockerKrbJDBCIntegrationSuite {
     override val jdbcPort = 3306
 
     override def getJdbcUrl(ip: String, port: Int): String =
-      s"jdbc:mysql://$ip:$port/mysql?user=$principal"
+      s"jdbc:mysql://$ip:$port/mysql?permitMysqlScheme&user=$principal"
```
Comment (panbingkun) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala:49]: permitMysqlScheme: 
https://mariadb.com/kb/en/mariadb-connector-j-303-release-notes/#jdbcmariadb-scheme

**Comment by panbingkun** [comment]
cc @yaooqinn

**Comment by yaooqinn** [comment]
shall we also update the server? https://mariadb.com/kb/en/mariadb-10-5-25-release-notes/

**Comment by panbingkun** [comment]
> shall we also update the server? https://mariadb.com/kb/en/mariadb-10-5-25-release-notes/

Sure, Let me verify locally first.

**Comment by panbingkun** [comment]
> shall we also update the server? https://mariadb.com/kb/en/mariadb-10-5-25-release-notes/

https://hub.docker.com/_/mariadb/tags?page=&page_size=&ordering=&name=10.5.25
It seems that the official image `10.5.25` is not ready yet, but `10.5.24` is ok.

**Review by yaooqinn — DISMISSED** [review]

File: connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MySQLIntegrationSuite.scala [hunk 2]
```diff
@@ -154,30 +154,43 @@ class MySQLIntegrationSuite extends DockerJDBCIntegrationSuite {
...
-    assert(rows.get(4).isInstanceOf[BigDecimal])
-    assert(rows.get(5).isInstanceOf[BigDecimal])
-    assert(rows.get(6).isInstanceOf[Double])
-    // Unlike MySQL, MariaDB seems not to distinguish signed and unsigned tinyint(1).
     val isMaria = jdbcUrl.indexOf("disableMariaDbDriver") == -1
     if (isMaria) {
+      assert(rows.get(0).isInstanceOf[Integer])
+      assert(rows.get(1).isInstanceOf[Long])
+      assert(rows.get(2).isInstanceOf[Integer])
+      assert(rows.get(3).isInstanceOf[BigDecimal])
```
Comment (dongjoon-hyun) [inline connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MySQLIntegrationSuite.scala:162]: Could you describe the root cause of this change, @panbingkun ? For me, this looks suspicious like a breaking change.
```scala
- assert(rows.get(3).isInstanceOf[Long])
+ assert(rows.get(3).isInstanceOf[BigDecimal])
```

Comment (panbingkun) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MySQLIntegrationSuite.scala:162]: the root cause of this change
- change:
  index[0], Short => Integer
  index[1, Integer => Long
  index[3], Long => Decimal

- Before - (mariadb jdbc driver: 2.7.12):
  image

- After - (mariadb jdbc driver: 3.4.0):
  image

Comment (panbingkun) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MySQLIntegrationSuite.scala:162]: We can also add compatible logic in `MySQLDialect##getCatalystType`. Do we need to do this in this PR?

Comment (panbingkun) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MySQLIntegrationSuite.scala:162]: We can make changes as follows to keep this compatible:
image

If we need to do this, how about making related changes directly in this PR? Or a new seperated PR?

Comment (yaooqinn) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MySQLIntegrationSuite.scala:162]: > After - (mariadb jdbc driver: 3.4.0):

These mariadb SQL Data Type <-> JDBC Data Type mapping rules seem incorrect, do they have a statement why they made such change

Comment (panbingkun) [reply connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MySQLIntegrationSuite.scala:162]: Let me try to find it.

**Comment by panbingkun** [comment]
Through comparative analysis, it is found that
1. In version `2.x`, there is `no distinction` between `signed` and `unsigned`. For example, a type is defined in MySQL: `med MEDIUMINT UNSIGNED`
  Through the `mariadb jdbc driver`, the final logic is:
  https://github.com/mariadb-corporation/mariadb-connector-j/blob/maintenance/2.x/src/main/java/org/mariadb/jdbc/MariaDbResultSetMetaData.java#L283
  image

  https://github.com/mariadb-corporation/mariadb-connector-j/blob/maintenance/2.x/src/main/java/org/mariadb/jdbc/internal/ColumnType.java#L73
  image
  ** `Types.INTEGER` will be returned  to `Spark` **

  On the Spark side, we will go to the following logic:
https://github.com/apache/spark/blob/14d3f447360b66663c8979a8cdb4c40c480a1e04/sql/core/src/main/scala/org/apache/spark/sql/jdbc/MySQLDialect.scala#L148-L151

  ** Therefore, `Interger` will be finally obtained. **

2. In version `3.x`, there is a `distinction` between `signed` and `unsigned`. For example, a type is defined in MySQL: `med MEDIUMINT UNSIGNED`
Through the `mariadb jdbc driver`, the final logic is:
  https://github.com/mariadb-corporation/mariadb-connector-j/blob/master/src/main/java/org/mariadb/jdbc/client/result/ResultSetMetaData.java#L232-L234
  image
  
  https://github.com/mariadb-corporation/mariadb-connector-j/blob/master/src/main/java/org/mariadb/jdbc/client/result/ResultSetMetaData.java#L297-L302
  image
  
  **UnsignedMediumIntColumn**
  https://github.com/mariadb-corporation/mariadb-connector-j/blob/master/src/main/java/org/mariadb/jdbc/client/column/UnsignedMediumIntColumn.java

**Comment by panbingkun** [comment]
IMO, it seems that the `3.x` version is more reasonable, but we are not compatible with this case in `Spark`, so I propose that we can make related changes on the `Spark side`

@yaooqinn @dongjoon-hyun

**Comment by yaooqinn** [comment]
> IMO, it seems that the 3.x version is more reasonable,

It's unreasonable to me. The 2.x and MySQL official both use `java.sql.Types.SMALLINT + Signed signal` = `[UN]SIGNED SMALLINT`

**Comment by panbingkun** [comment]
> > IMO, it seems that the 3.x version is more reasonable,
> 
> It's unreasonable to me. The 2.x and MySQL official both use `java.sql.Types.SMALLINT + Signed signal` = `[UN]SIGNED SMALLINT`

Okay, I see what you mean. What you say is `reasonable` is that you focus on the value returned in `3.x` and do not make full use of `isSigned`. What I say is `reasonable` is that you focus on its own division of `UnsignedSmallIntColumn` and `SignedSmallIntColumn` without paying attention to its `isSigned`. Let me investigate again.
it should have been like `this` in the original version of `3.x` (from 3.0.3)
https://github.com/mariadb-corporation/mariadb-connector-j/commit/22cff70de2be679e925a0d231123ad72f30ffdd6

**Comment by panbingkun** [comment]
Regarding the above issues, I have proposed a PR to `mariadb-connector-j`: https://github.com/mariadb-corporation/mariadb-connector-j/pull/194

**Pull request closed without merging**